mlir/lib/Conversion/MemRefToSPIRV/MapMemRefStorageClassPass.cpp

```diff
@@ -335,23 +287,25 @@ class MapMemRefStorageClassPass final
     MLIRContext *context = &getContext();
     Operation *op = getOperation();
 
+    spirv::MemorySpaceToStorageClassMap spaceToStorage = memorySpaceMap;
     if (spirv::TargetEnvAttr attr = spirv::lookupTargetEnv(op)) {
       spirv::TargetEnv targetEnv(attr);
       if (targetEnv.allows(spirv::Capability::Kernel)) {
-        memorySpaceMap = spirv::mapMemorySpaceToOpenCLStorageClass;
+        spaceToStorage = spirv::mapMemorySpaceToOpenCLStorageClass;
       } else if (targetEnv.allows(spirv::Capability::Shader)) {
-        memorySpaceMap = spirv::mapMemorySpaceToVulkanStorageClass;
+        spaceToStorage = spirv::mapMemorySpaceToVulkanStorageClass;
       }
     }
 
-    std::unique_ptr<ConversionTarget> target =
-        spirv::getMemorySpaceToStorageClassTarget(*context);
-    spirv::MemorySpaceToStorageClassConverter converter(memorySpaceMap);
-
-    RewritePatternSet patterns(context);
-    spirv::populateMemorySpaceToStorageClassPatterns(converter, patterns);
+    spirv::MemorySpaceToStorageClassConverter converter(spaceToStorage);
+    // Perform the replacement.
+    spirv::convertMemRefTypesAndAttrs(op, converter);
 
-    if (failed(applyFullConversion(op, *target, std::move(patterns))))
+    // Only perform the conversion to check that there are no illegal ops
+    // remaining. Do not attempt to convert anything.
+    if (failed(applyFullConversion(
+            op, *spirv::getMemorySpaceToStorageClassTarget(*context),
+            RewritePatternSet{context})))
```

Comment:
I'm not an expert on dialect conversion, but does this just boil down to a walk + ConversionTarget::isLegal? Ditto for the above.

Reply:
Yes, exactly. But there's no better named function for this AFAIK. We could simplify this but not exposing a conversion target at all and expect the downstream users to walk this manually.

I'm not sure which one is better honestly -- if we save on some complexity here, we increase complexity downstream.

Reply:
Can we keep the conversion target and just in these places do
```
op->walk([&](Operation *target) {
  if (!conversionTarget.isLegal(target)) {
    return WalkResult::interrupt();
  }
});
```
?

Reply:
Yeah an explicit walk looks better to me--it's simpler and clearer so easy to understand. also we avoid the overhead of conversion framework.

Reply:
OK, I will change this to a walk. But I don't see any complexity in the conversion framework -- it effectively does the same thing.